task/google/resources/resource_instance_template.go

```diff
@@ -0,0 +1,226 @@
+	"time"
+
+	"google.golang.org/api/compute/v1"
+	"google.golang.org/api/googleapi"
+
+	"terraform-provider-iterative/task/google/client"
+	"terraform-provider-iterative/task/universal"
+)
+
+func NewInstanceTemplate(client *client.Client, identifier string, defaultNetwork *DefaultNetwork, firewallRules []*FirewallRule, image *Image, credentials *Credentials, task universal.Task) *InstanceTemplate {
+	i := new(InstanceTemplate)
+	i.Client = client
+	i.Identifier = universal.NormalizeIdentifier(identifier, true)
+	i.Attributes = task
+	i.Dependencies.Credentials = credentials
+	i.Dependencies.DefaultNetwork = defaultNetwork
+	i.Dependencies.FirewallRules = firewallRules
+	i.Dependencies.Image = image
+	return i
+}
+
+type InstanceTemplate struct {
+	Client       *client.Client
+	Identifier   string
+	Attributes   universal.Task
+	Dependencies struct {
+		*DefaultNetwork
+		FirewallRules []*FirewallRule
+		*Image
+		*Credentials
+	}
+	Resource *compute.InstanceTemplate
+}
+
+func (i *InstanceTemplate) Create(ctx context.Context) error {
+	keyPair, err := i.Client.GetKeyPair(ctx)
+	if err != nil {
+		return err
+	}
+
+	publicKey, err := keyPair.PublicString()
+	if err != nil {
+		return err
+	}
+
+	formattedPublicKey := fmt.Sprintf("%s:%s host\n", i.Dependencies.Image.Attributes.SSHUser, strings.TrimSpace(publicKey))
+
+	if i.Attributes.Environment.Variables == nil {
+		i.Attributes.Environment.Variables = make(map[string]*string)
+	}
+	for name, value := range *i.Dependencies.Credentials.Resource {
+		valueCopy := value
+		i.Attributes.Environment.Variables[name] = &valueCopy
+	}
+
+	script := i.Attributes.Environment.GenerateMachineScript()
+
+	size := i.Attributes.Size.Machine
+	sizes := map[string]string{
+		"m":       "e2-custom-8-32768",
+		"l":       "e2-custom-32-131072",
+		"xl":      "n2-custom-64-262144",
+		"m+k80":   "custom-8-53248+nvidia-tesla-k80*1",
+		"l+k80":   "custom-32-131072+nvidia-tesla-k80*4",
+		"xl+k80":  "custom-64-212992-ext+nvidia-tesla-k80*8",
+		"m+v100":  "custom-8-65536-ext+nvidia-tesla-v100*1",
+		"l+v100":  "custom-32-262144-ext+nvidia-tesla-v100*4",
+		"xl+v100": "custom-64-524288-ext+nvidia-tesla-v100*8",
+	}
+
+	if val, ok := sizes[size]; ok {
+		size = val
+	}
+
+	accelerators := []*compute.AcceleratorConfig{}
+
+	match := regexp.MustCompile(`^([^+]+)(?:\+([^*]+)\*([1-9]\d*))?$`).FindStringSubmatch(size)
+	if match == nil {
+		return errors.New("invalid machine type")
+	}
+
+	machineType := match[1]
+	if match[2] != "" {
+		acceleratorType := match[2]
+		acceleratorCount, err := strconv.Atoi(match[3])
+		if err != nil {
+			return err
+		}
+		accelerators = append(accelerators, &compute.AcceleratorConfig{
+			AcceleratorCount: int64(acceleratorCount),
+			AcceleratorType:  acceleratorType,
+		})
+	}
+
+	var isPreemptible bool
+	switch i.Attributes.Spot {
+	case 0:
+		isPreemptible = true
+	case -1:
+		isPreemptible = false
+	default:
+		return errors.New("preemptible instances don't have bidding price")
+	}
```

Comment:
```suggestion
if i.Attributes.Spot > 0 {
        return errors.New("preemptible instances don't have bidding price")
}
isPreemptible := i.Attributes.Spot == 0
```

Reply:
if i.Attributes.Spot != 0 && i.Attributes.Spot !=-1 {
        return errors.New("preemptible instances don't have bidding price")
}
isPreemptible := i.Attributes.Spot == 0

just to be very exact ;)

Reply:
Truth be told, other providers accept any negative value, so let's consider any negative value == disabled.

Reply:
I was too purist on this conversation. 🙈